### PR TITLE
CA-90

### PR DESCRIPTION
**Pull request by JongChern**
add a class to support multiple byte reading

### CODE REVIEW

File: ShimmerAPI/ShimmerAPI/ShimmerLogAndStreamSystemSerialPortV2.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                        }
+                        else
+                        {
+                            if (ReadStopWatch.ElapsedMilliseconds > ReadTimeout) //If no bytes have been read throw an exception
+                            {
+                                ReadStopWatch.Restart();
+                                throw new System.TimeoutException();
+                            }
+                        }
+                        Thread.Sleep(10);
```
Comment (davidracl) [inline ShimmerAPI/ShimmerAPI/ShimmerLogAndStreamSystemSerialPortV2.cs:59]: You should rewrite it to Async Task and use await Task.Delay. In this case you block the current thread at all.

**Review by davidracl — CHANGES_REQUESTED** [review]

**Comment by jarekkonecny** [comment]
Dear all,

I am the original author of this improvement and I feel that the point of proposed code is not well-understand. The main point is a computational cost optimization.
ShimmerApi uses SerialPort.ReadByte() method to read data from serial port. However, this method has the highest computational cost from all available methods.
Let us assume this piece of the code:

```
void readThread()
{
    while(true) 
    {
        int mbyte = sp.Read();
        if (mbyte >= 0) 
        {
            // Process byte
        }
        else
        {
            // No byte received
            Thread.Sleep(10); // Sleep for some time is required, otherwise CPU load is very high
        }
    }
}
```
This piece of code is not optimal for reading, because sp.Read() is slow.
We can change the code this way in order to make it faster:

```
void readThread()
{
    while(true) 
    {
    int bytesToRead = sp.BytesToRead; // How many bytes are received? Copy this value to some variable, because it can change anytime.
        if (bytesToRead > 0) 
        {
            byte[] buffer = new byte[bytesToRead];
            sp.Read(buffer, 0, bytesToRead);
            for(i = 0; i < bytesToRead; i++) 
            {
                byte mbyte = buffer[i];
                // Process byte
            }
        }
        else
        {
            // No byte received
            Thread.Sleep(10); // Sleep for some time is required, otherwise CPU load is very high
        }
    }
}
```

The codes demonstrates the original point. ShimmerApi uses different architecture to read data. We override ShimmerApi ReadByte method and we created a finite state machine, that emulates single byte reading, although multiple bytes are read and stored to a memory stream.
My opinion is that the proposed code in the pull request significantly optimize the computational costs, however, this particular implementation is not written in clean way.

**Comment by JongChern** [comment]
@jarekkonecny thanks, we will accept the PR since it serves the purpose of making it more efficient, we can always create a new PR if we want to improve the code. I myself don't have the time for this unfortunately. Thanks for your contributions.

**Review by weiwentan23 — APPROVED** [review]
Looks good

**Pull request merged**